08.ls_object/ls_app.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+require_relative './ls_file'
+
+class LsApp
+  COLUMNS_COUNT = 3
+  MARGIN_COUNT = 20
+
+  def initialize(input_options)
+    @input_options = input_options
+    @ls_files = generate_ls_files
+    @ls_files = @ls_files.reverse if @input_options['r']
+  end
+
+  def run
+    if @input_options['l']
+      display_long
+    else
+      display_short
+    end
+  end
+
+  private
+
+  def generate_ls_files
+    file_names = 
+      if @input_options['a']
+        Dir.foreach('.')
+      else
+        Dir.glob('*')
+      end
+    file_names = file_names.map { |file_name| LsFile.new(file_name) }
+    file_names.sort_by{|f| f.name }
+  end
+
+  def display_long
+    puts "total #{@ls_files.map(&:blocks).sum}"
+    nlink_count = max_nlink_count
+    owner_count = max_owner_count
+    group_count = max_group_count
+    ls_file_size_count = max_ls_file_size_count
+
+    @ls_files.each do |ls_file|
+      output = []
+      output << "#{ls_file.type}#{ls_file.permission} "
+      output << ls_file.nlink.to_s.rjust(nlink_count)
+      output << "#{ls_file.owner.to_s.ljust(owner_count)} "
+      output << "#{ls_file.group.to_s.ljust(group_count)} "
+      output << ls_file.size.to_s.rjust(ls_file_size_count)
+      output << ls_file.modified_time.to_s
+      output << ls_file.name.to_s
+      puts output.join(' ')
+    end
+  end
+
+  def display_short
+    file_names = []
+    all_files = []
+    @ls_files.each_with_index do |file, i|
+      file_names << file.name.ljust(MARGIN_COUNT)
+      if ((i + 1) % COLUMNS_COUNT).zero?
+        all_files << file_names
+        file_names = []
+        all_files
+      end
+    end
+    all_files << file_names unless file_names.empty?
+
+    # 行列を入れ替えるため最後の配列に足りない要素数を加えています。
+    if all_files.last.count != all_files.first.count
+      (all_files.first.count - all_files.last.count).times { all_files.last << [ ] }
+    end
+    all_files = all_files.transpose
+    all_files.each { |file| puts file.join }
+  end
+
+  def max_group_count
+    group_count = @ls_files.map { |file| file.group.to_s.split('').size }
+    group_count.max
```

Comment:
```suggestion
    @ls_files.map { |file| file.group.to_s.split('').size }.max
```

だけでも良さそうです。

あと`split.size`のように`('')`を省略するのもアリだと思いますー。

Reply:
こちらも内容理解しました(__)
修正します。

Reply:
@JunichiIto 
split('')なしだと正しい出力にならなかったので、一旦つけたままプッシュしました。
[3132761](https://github.com/yuki-snow1823/ruby-practices/pull/2/commits/313276178008b0c16329bdfbea93fe869dd25a3c)

Reply:
あ、コメントを間違えたかも。`split('')` == `chars` ですよ、って言いたかったんだと思います。どうもすいません 🙏 
ちなみに配列にする必要がない（１文字ずつループさせるだけで良い）ときは`each_char`も使えます。
詳しくはAPIドキュメントを参照してください〜。

Reply:
>chars -> [String][[permalink](https://docs.ruby-lang.org/ja/latest/method/String/i/chars.html)][[rdoc](https://docs.ruby-lang.org/en/3.1/String.html#method-i-chars)][[edit](https://github.com/rurema/doctree/edit/master/refm/api/src/_builtin/String#L3340)]
chars {|cstr| block } -> self
文字列の各文字を文字列の配列で返します。(self.each_char.to_a と同じです)

こんなものがあったとは...すぐに思いついたロジックで実装する前に、先にこういったものがすでにないか調べようと思いました。
https://docs.ruby-lang.org/ja/latest/method/String/i/chars.html

こちらで対応しました。
https://github.com/yuki-snow1823/ruby-practices/pull/2/commits/9134698aabe4dd5022daf9353cba613bf943a99f